### PR TITLE
Test for SBERT

### PR DESCRIPTION
**Pull request by mmaksimovicTT**
Added a test for the SBERT model.

### CODE REVIEW

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    # @override
+    def _get_forward_method_name(self) -> str:
+        return "apply"
+
+    # @override
+    def _get_input_activations(self) -> Dict[str, jax.Array]:
+        tokenizer = AutoTokenizer.from_pretrained(MODEL_PATH)
+        inputs = tokenizer(
+            ["Bu örnek bir cümle", "Her cümle vektöre çevriliyor"],
```
Comment (kmitrovicTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:62]: Leave a comment that this model was trained on Turkish sentences, just so it's clear.

**Review by kmitrovicTT — APPROVED** [review]

File: tests/jax/single_chip/models/sbert/model_implementation.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+# SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+from flax import linen as nn
+import jax.numpy as jnp
+from transformers import FlaxBertModel, BertConfig
+
+
+class FlaxSentenceTransformerBERT(nn.Module):
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/model_implementation.py:9]: As documented [here](https://github.com/UKPLab/sentence-transformers/blob/master/docs/sentence_transformer/usage/custom_models.rst#structure-of-sentence-transformer-models), Sentence Transformer model is formed as a combination of these components:
- Pretrained Transformer model
- Pooling layer
- Dense layer (optional)
- Normalize layer (optional)

You can find which components model uses by checking the `modules.json` file in the model HF package. For example in SBERT variant Dense layer and normalization is not done, can be seen [here](https://huggingface.co/emrecan/bert-base-turkish-cased-mean-nli-stsb-tr/blob/main/modules.json) (there is no dense or normalization modules listed).

Since you've already added implementation for some of the things that are not needed for the SBERT variant (like prompt masking and normalization), I would keep them and rename this class to `FlaxSentenceTransformer` class, and then create a separate `FlaxSentenceTransformerBERT` class which extends it and initializes things according to SBERT specification.

Here is how that could look like:
```
class PoolingMode(Enum):
    # Use the first token (CLS token) as text representations.
    CLS_TOKEN = "cls_token"

    # Use max in each dimension over all tokens.
    MAX_TOKENS = "max_tokens"

    # Perform mean-pooling.
    MEAN_TOKENS = "mean_tokens"

    # Perform mean-pooling, but divide by sqrt(input_length).
    MEAN_SQRT_LEN_TOKENS = "mean_sqrt_len_tokens"

    # Perform (position) weighted mean pooling (https://arxiv.org/abs/2202.08904).
    WEIGHTEDMEAN_TOKENS = "weightedmean_tokens"

    # Perform last token pooling (https://arxiv.org/abs/2201.10005).
    POOLING_MODE_LASTTOKEN = "pooling_mode_lasttoken"


class FlaxSentenceTransformer(nn.Module):
    """
    Sentence Transformer model is formed as a combination of these components:
        - Pretrained Transformer model
        - Pooling layer
        - Dense layer (optional)
        - Normalize layer (optional)
    """

    transformer: nn.Module
    pooling_mode: PoolingMode
    dense_layer: nn.Module = None
    normalize: bool = False

    # Some models (INSTRUCTOR, NV-Embed) exclude the prompt from the mean pooling step,
    # such that it's only used in the Transformer blocks.
    pooling_include_prompt: bool = True
    prompt_length: int = 0

    def get_params(self) -> PyTree:
        return self.transformer.params

    @nn.compact
    def __call__(
        self,
        input_ids: jnp.ndarray,
        params: dict,
        attention_mask: jnp.ndarray,
        token_type_ids: jnp.ndarray = None,
    ) -> jnp.ndarray:
        transformer_outputs = self.transformer(
            input_ids=input_ids,
            params=params,
            attention_mask=attention_mask,
            token_type_ids=token_type_ids,
        )
        token_embeddings = transformer_outputs.last_hidden_state

        # Mean pooling
        if not self.pooling_include_prompt and self.prompt_length > 0:
            attention_mask = attention_mask.at[:, :self.prompt_length].set(0)
        assert self.pooling_mode == PoolingMode.MEAN_TOKENS, "Only mean pooling supported for now"
        if self.pooling_mode == PoolingMode.MEAN_TOKENS:
            input_mask_expanded = jnp.expand_dims(attention_mask, axis=-1)
            sum_embeddings = jnp.sum(token_embeddings * input_mask_expanded, axis=1)
            sum_mask = jnp.clip(jnp.sum(input_mask_expanded, axis=1), a_min=1e-9)
            embeddings = sum_embeddings / sum_mask

        # TODO Add support for dense layer.
        assert self.dense_layer == None, "Dense layer is not supported yet"

        if self.normalize:
            norm = jnp.linalg.norm(embeddings, axis=1, keepdims=True)
            embeddings = embeddings / norm

        return embeddings


class FlaxSentenceTransformerBERT(nn.Module):
    def __init__(
        self,
        pretrained_model: nn.Module,
    ) -> None:
        super().__init__(
            transformer = pretrained_model,
            pooling_mode = PoolingMode.MEAN_TOKENS,
            )
```

File: tests/jax/single_chip/models/sbert/model_implementation.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        self,
+        input_ids: jnp.ndarray,
+        attention_mask: jnp.ndarray,
+        token_type_ids: jnp.ndarray = None,
+        normalize: bool = True,
+        include_prompt: bool = True,
+        prompt_length: int = 0,
+    ) -> jnp.ndarray:
+        if not include_prompt and prompt_length > 0:
+            attention_mask = attention_mask.at[:, :prompt_length].set(0)
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/model_implementation.py:23]: As mentioned [here](https://github.com/UKPLab/sentence-transformers/blob/dd8a8d7d4d5660e7b2b69e8a1ece8202417c0069/examples/sentence_transformer/training/prompts/README.md?plain=1#L91) and [here](https://github.com/UKPLab/sentence-transformers/blob/dd8a8d7d4d5660e7b2b69e8a1ece8202417c0069/sentence_transformers/models/Pooling.py#L37), this is done only for the pooling component. In the suggested code I've provided in the comment above I've moved it to happen right before the pooling.

File: tests/jax/single_chip/models/sbert/model_implementation.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+        model = FlaxBertModel(self.config, dtype=jnp.float32)
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/model_implementation.py:25]: This initializes model with the random parameters, which is probably why you are getting the `Failed to legalize operation 'stablehlo.shift_right_logical'` error, because the `__call__` is executed on device and parameters initialization uses RNG, and we do not support RNG operations on device yet.

You should use the loaded pretrained BERT model instead. In the suggested code I've provided in the comment above I've added `pretrained_model` parameter to the `FlaxSentenceTransformerBERT ` class, you should set that parameter from the tester in `_get_model` function, I'll leave the comment there too.

File: tests/jax/single_chip/models/sbert/model_implementation.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        output = model(
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/model_implementation.py:26]: You need to pass the loaded model params here, otherwise it will use `self.params` which is set only for the randomly initialized params like when model is created from the config, like you did above.

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+        super().__init__(comparison_config, run_mode)
+
+    # @override
+    def _get_model(self) -> nn.Module:
+        config = BertConfig.from_pretrained(MODEL_PATH)
+        return FlaxSentenceTransformerBERT(config=config)
+
+    # @override
+    def _get_forward_method_name(self) -> str:
+        return "apply"
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:56]: Can be removed, our `FlaxSentenceTransformer` class is using `__call__` method which is the default name on model tester.

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        comparison_config: ComparisonConfig = ComparisonConfig(),
+        run_mode: RunMode = RunMode.INFERENCE,
+    ) -> None:
+        self._model_path = model_path
+        super().__init__(comparison_config, run_mode)
+
+    # @override
+    def _get_model(self) -> nn.Module:
+        config = BertConfig.from_pretrained(MODEL_PATH)
+        return FlaxSentenceTransformerBERT(config=config)
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:52]: This now has to be changed to load the pretrained BERT model and pass it to the `FlaxSentenceTransformerBERT` constructor. Also should use `self._model_path` instead of `MODEL_PATH`.

Like this:
```
pretrained_model = FlaxBertModel.from_pretrained(self._model_path, from_pt=True)
return FlaxSentenceTransformerBERT(pretrained_model)
```

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+            return_tensors="jax",
+        )
+        return inputs
+
+    # @override
+    def _get_forward_method_kwargs(self) -> Dict[str, jax.Array]:
+        pretrained_model = FlaxBertModel.from_pretrained(MODEL_PATH, from_pt=True)
+        params = pretrained_model.params
+
+        inputs = self._get_input_activations()
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:77]: These lines can now be removed, and below you will use:
```
return {
    "params": self._input_parameters,
    "input_ids": self._input_activations["input_ids"],
    "attention_mask": self._input_activations["attention_mask"],
    "token_type_ids": self._input_activations.get("token_type_ids"),
}
```

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+            "attention_mask": inputs["attention_mask"],
+            "token_type_ids": inputs.get("token_type_ids"),
+            "normalize": True,
+            "include_prompt": True,
+            "prompt_length": 0,
+        }
+
+    # @override
+    def _get_static_argnames(self):
+        return ["normalize", "include_prompt"]
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:90]: Not needed anymore

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    model_name=MODEL_NAME,
+    model_group=ModelGroup.RED,
+    run_mode=RunMode.INFERENCE,
+    bringup_status=BringupStatus.FAILED_FE_COMPILATION,
+)
+@pytest.mark.xfail(
+    reason=failed_fe_compilation(
+        "Failed to legalize operation 'stablehlo.shift_right_logical'. "
+        "https://github.com/tenstorrent/tt-xla/issues/417"
+    )
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:122]: Should be updated now

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+        run_mode: RunMode = RunMode.INFERENCE,
+    ) -> None:
+        self._model_path = model_path
+        super().__init__(comparison_config, run_mode)
+
+    # @override
+    def _get_model(self) -> nn.Module:
+        config = BertConfig.from_pretrained(MODEL_PATH)
+        return FlaxSentenceTransformerBERT(config=config)
+
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:53]: You now have to override the `def _get_input_parameters(self) -> PyTree:` method, and add this implementation:
```
# @override
def _get_input_parameters(self) -> PyTree:
    return self._model.get_params()
```

File: tests/jax/single_chip/models/sbert/test_sbert.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+        tokenizer = AutoTokenizer.from_pretrained(MODEL_PATH)
+
+        # The model weights used in this test were trained on Turkish sentence pairs.
+        # These sentences below match the training data distribution.
+        inputs = tokenizer(
+            ["Bu örnek bir cümle", "Her cümle vektöre çevriliyor"],
+            padding=True,
+            truncation=True,
+            return_tensors="jax",
+        )
```
Comment (mrakitaTT) [inline tests/jax/single_chip/models/sbert/test_sbert.py:69]: In Sentence Transformer they are calling tokenizer with specific configuration, as can be seen [here](https://github.com/UKPLab/sentence-transformers/blob/dd8a8d7d4d5660e7b2b69e8a1ece8202417c0069/sentence_transformers/models/Transformer.py#L507). They are using `truncation="longest_first"` and `max_length=self.max_seq_length`. Parameter `max_seq_length` is loaded from the HF model package file `sentence_bert_config.json` [here](https://huggingface.co/emrecan/bert-base-turkish-cased-mean-nli-stsb-tr/blob/main/sentence_bert_config.json). Ideally we would parse this info from the config file, but for now we can just hardcode the value into the constant here before the tokenizer call, with a todo comment to do it later.
